sph/include/sph/particles_data.hpp

```diff
@@ -123,12 +132,19 @@ class ParticlesData
      */
     auto data()
     {
-        using FieldType = std::variant<std::vector<float>*, std::vector<double>*, std::vector<int>*>;
+        using IntVecType     = std::decay_t<decltype(neighborsCount)>;
+        using KeyVecType     = std::decay_t<decltype(codes)>;
+        using FieldAllocType = typename std::decay_t<decltype(x)>::allocator_type;
+        using FieldType      = std::variant<std::vector<float, FieldAllocType>*,
+                                       std::vector<double, FieldAllocType>*,
+                                       KeyVecType*,
+                                       IntVecType*>;
 
         std::array<FieldType, fieldNames.size()> ret{
-            &x,   &y,   &z,        &x_m1,     &y_m1,     &z_m1, &vx,    &vy,    &vz,    &rho,  &u,     &p,     &h,
-            &m,   &c,   &grad_P_x, &grad_P_y, &grad_P_z, &du,   &du_m1, &dt,    &dt_m1, &c11,  &c12,   &c13,   &c22,
-            &c23, &c33, &mue,      &mui,      &temp,     &cv,   &rho0,  &wrho0, &kx,    &divv, &curlv, &alpha, &gradh};
+            &x,     &y,     &z,     &x_m1,     &y_m1,          &z_m1,     &vx,   &vy,    &vz,   &rho,   &u,   &p,
+            &h,     &m,     &c,     &grad_P_x, &grad_P_y,      &grad_P_z, &du,   &du_m1, &dt,   &dt_m1, &c11, &c12,
+            &c13,   &c22,   &c23,   &c33,      &mue,           &mui,      &temp, &cv,    &rho0, &wrho0, &kx,  &divv,
+            &curlv, &alpha, &gradh, &codes,    &neighborsCount};
```

Comment:
What are "codes"? The SFC keys? Why are they needed as a field?

Reply:
yes, these are the SFC keys. The are needed for tree-building and for neighbor particle searches.

Reply:
The "codes" is quite old, and is being replaced by `particleKeys`. The comment in the field definitions mentions though what it is:
```c++
std::vector<KeyType> codes;                        // Particle space-filling-curve keys
```